huaweicloud/services/smn/resource_huaweicloud_smn_topic.go

```diff
@@ -51,117 +74,180 @@ func ResourceTopic() *schema.Resource {
+	log.Printf("[DEBUG] Create Options: %#v", createOpts)
 
 	topic, err := topics.Create(client, createOpts).Extract()
 	if err != nil {
-		return fmtp.Errorf("Error getting topic from result: %s", err)
+		return diag.Errorf("error getting SMN topic from result: %s", err)
 	}
-	logp.Printf("[DEBUG] Create : topic.TopicUrn %s", topic.TopicUrn)
-	if topic.TopicUrn != "" {
-		d.SetId(topic.TopicUrn)
-		return resourceTopicRead(d, meta)
+	log.Printf("[DEBUG] Successfully created SMN topic: %s", topic.TopicUrn)
+
+	d.SetId(topic.TopicUrn)
+
+	//set tags
+	tagRaw := d.Get("tags").(map[string]interface{})
+	if len(tagRaw) > 0 {
+		taglist := utils.ExpandResourceTags(tagRaw)
+		tagClient, err := config.SmnV2TagClient(config.GetRegion(d))
+		if err != nil {
+			return diag.Errorf("error creating SMN tag client: %s", err)
+		}
+		tagClient.MoreHeaders = map[string]string{
+			"X-SMN-RESOURCEID-TYPE": "name",
+		}
+		if tagErr := tags.Create(tagClient, "smn_topic", d.Get("name").(string), taglist).ExtractErr(); tagErr != nil {
+			return diag.Errorf("error setting tags of SMN topic %s: %s", d.Id(), tagErr)
+		}
 	}
 
-	return fmtp.Errorf("Unexpected conversion error in resourceTopicCreate.")
+	return resourceTopicRead(ctx, d, meta)
 }
 
-func resourceTopicRead(d *schema.ResourceData, meta interface{}) error {
+func resourceTopicRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	config := meta.(*config.Config)
 	client, err := config.SmnV2Client(config.GetRegion(d))
 	if err != nil {
-		return fmtp.Errorf("Error creating HuaweiCloud smn client: %s", err)
+		return diag.Errorf("error creating SMN client: %s", err)
 	}
 
 	topicUrn := d.Id()
 	topicGet, err := topics.Get(client, topicUrn).ExtractGet()
 	if err != nil {
-		return common.CheckDeleted(d, err, "topic")
+		return common.CheckDeletedDiag(d, err, "error retrieving SMN topic")
 	}
 
-	logp.Printf("[DEBUG] Retrieved topic %s: %#v", topicUrn, topicGet)
+	log.Printf("[DEBUG] Retrieved SMN topic %s: %#v", topicUrn, topicGet)
+
+	mErr := multierror.Append(
+		d.Set("region", config.GetRegion(d)),
+		d.Set("topic_urn", topicGet.TopicUrn),
+		d.Set("display_name", topicGet.DisplayName),
+		d.Set("name", topicGet.Name),
+		d.Set("push_policy", topicGet.PushPolicy),
+		d.Set("update_time", topicGet.UpdateTime),
+		d.Set("create_time", topicGet.CreateTime),
+		d.Set("enterprise_project_id", topicGet.EnterpriseProjectId),
+	)
+
+	// fetch tags
+	tagClient, err := config.SmnV2TagClient(config.GetRegion(d))
+	if err != nil {
+		return diag.Errorf("error creating SMN tag client: %s", err)
+	}
+	tagClient.MoreHeaders = map[string]string{
+		"X-SMN-RESOURCEID-TYPE": "name",
+	}
+	if resourceTags, err := tags.Get(tagClient, "smn_topic", d.Get("name").(string)).Extract(); err == nil {
+		tagmap := utils.TagsToMap(resourceTags.Tags)
+		mErr = multierror.Append(mErr, d.Set("tags", tagmap))
+	} else {
+		log.Printf("[WARN] fetching tags of SMN topic failed: %s", err)
```

Comment:
seems `return diag.Errorf()`?

Reply:
api of tags not available in some region